### ISSUE TITLE
Deterministic Annotations

### ISSUE THREAD

**Issue by o-o00o-o** [otykier/TabularEditor#217]
Hi, so now we have a bunch of data dictionary descriptions etc in our tables I have changed the serialisation method so that Annotations and Partitions are in the table.json file as well as putting relationships by their starting tables.

While doing a few imports and fixing things I found that the TabularEditor_Relationships annotation jumped from one position to another 

The new file looks like it is now sorting the annotations into order correctly and possibly this was something that was fixed in the last version to make it deterministic but if not just a heads up.

Previously TabularEditor_RElationships was at the bottom where now with the new version from today it is putting it after TabularEditor_InPerspective and before my VR_* annotations

**Comment by otykier** [comment]
Annotations are not deterministic at the moment, so this is a good point. I'll make sure that they are sorted alphabetically by annotation name going forward.

**Comment by otykier** [comment]
On second thoughts - perhaps this should be an optional setting. By default, TOM arranges objects in the JSON in the order they are inserted, so suddenly forcing alphabetical ordering in the JSON files might cause a lot of unexpected diffs for people. And this goes for all object types - not just annotations.

**Comment by o-o00o-o** [comment]
That is true probably wise to have a control to allow people to switch when they want but default for new models. You really want this to be deterministic otherwise you get strange changes as a result of scripts running that basically produce the same logical result but git sees it as a change

**Comment by otykier** [comment]
We need to think carefully about this. Today, a simple change such as cutting a table away (Ctrl+X) and subsequently pasting it back in (Ctrl+V) would be enough to change the current ordering within the metadata. Interestingly, noone has complained about that yet. Forcing alphabetical order would of course prevent a change of ordering in this case.

On the other hand, alphabetical order might cause a change of ordering in cases when objects are renamed... I think that might actually be worse, since that probably happens more often.

**Comment by greggyb** [comment]
I think there is still room for improvement here. I often find that annotations are re-ordered in objects that I have made no changes to. For example, just now I deleted several columns from my model. When I save it out, I am seeing annotation re-ordering in objects I haven't touched.

Specifically, I deleted `Table1[ColumnA]`, and `Table2[MeasureB]` has its annotations re-ordered. The only relationship between these things is that they exist in the same perspectives, but I can't confirm that this is the causal factor.

It seems that there is room for a stricter preservation of existing ordering of annotations in existing objects, regardless of whether there is any capability to enforce ordering.

Here's a sample from my diff (scrubbed names):

I see this sort of thing on a semi-regular basis (maybe around new releases of TE? I did just update today)

```diff
--- a/path/cube/tables/table1/measures/measure.json
+++ b/path/cube/tables/table1/measures/measure.json
@@ -34,13 +34,13 @@
         ")"
       ]
     },
-    {
-      "name": "TabularEditor_InPerspective",
-      "value": "[\"persp1\",\"persp2\",\"persp3\"]"
-    },
     {
       "name": "persp2_Expression",
       "value": "[MeasureReference]"
+    },
+    {
+      "name": "TabularEditor_InPerspective",
+      "value": "[\"persp1\",\"persp2\",\"persp3\"]"
     }
   ]
 }
```

This is at least partially deterministic, because it will continue to occur when I make similar unrelated changes. Thus I am forced to either continuously exclude changes introduced by TE, or accept them and muddy up my version history. While BISM Normalizer is a great tool, it is not a good solution to use as part of a normal development cycle. I want to be able to `git diff` and see changes. Setting my difftool to BISM Normalizer would introduce integer-minute delays into a workflow that typically takes seconds currently.
